<launch>
   <arg name="multi_robot"             default="robot1"/>
   <arg name="use_rviz"                default="true"/>
   <arg name="use_bag"                 default="true"/>
   <arg name="use_remote"              default="false"/>
   <arg name="topic_cmd"               default="cmd_vel"/>
   <arg name="topic_odom"              default="/robot1/odom/noise"/>
   <arg name="topic_rgb"               default="/camera/color/image_raw"/>
   <arg name="topic_depth"             default="/camera/aligned_depth_to_color/image_raw"/>
   <arg name="topic_info"              default="/camera/color/camera_info"/>
   <arg name="frame_id_odom"           default="$(arg multi_robot)/odom"/>
   <arg name="frame_id_robot_base"     default="$(arg multi_robot)//base_footprint"/>
   <arg name="frame_id_camera_link"    default="camera_link"/>
   <arg name="file_rviz"               default="$(find vmcl)/rviz/show_particles.rviz"/>
   <arg name="file_marker_pose"        default="$(find vmcl)/config/aruco_marker.yaml"/>
   <arg name="file_urdf"               default="$(find turtlebot3_description)/urdf/turtlebot3_waffle.urdf.xacro"/>

   <group if="$(arg use_remote)">
      <include file="$(find potbot_example)/launch/env/remote.launch">
        <arg name="urdf_file"               value="$(find xacro)/xacro --inorder '$(arg file_urdf)'"/>
        <arg name="multi_robot_name"        value="$(arg multi_robot)"/>
      </include>
    </group>

   <node pkg="vmcl" name="vmcl" type="vmcl" output="screen">
      <remap from="odom"               to="$(arg topic_odom)"/>
      <remap from="color/image_raw"    to="$(arg topic_rgb)"/>
      <remap from="depth/image_raw"    to="$(arg topic_depth)"/>
      <remap from="color/camera_info"  to="$(arg topic_info)"/>

      <param name="frame_id_camera_link"                 value="$(arg frame_id_camera_link)"/>

      <param name="norm_noise_mean_linear_linear"        value="0.0"/>
      <param name="norm_noise_variance_linear_linear"    value="0.0005"/>
      <param name="norm_noise_mean_linear_angular"       value="0.0"/>
      <param name="norm_noise_variance_linear_angular"   value="0.0005"/>
      <param name="norm_noise_mean_angular_linear"       value="0.0"/>
      <param name="norm_noise_variance_angular_linear"   value="0.001"/>
      <param name="norm_noise_mean_angular_angular"      value="0.0"/>
      <param name="norm_noise_variance_angular_angular"  value="0.005"/>
      <param name="depth_scaling"                        value="1000"/>
      <param name="move_mean_window_num"                 value="40"/>
   </node>

   <node pkg="vmcl" name="marker_broadcaster" type="marker_broadcaster" output="screen">
      <rosparam file="$(arg file_marker_pose)" command="load" />
   </node>
   
   <group if="$(arg use_bag)">
      <!-- <node pkg="tf" type="static_transform_publisher" name="robot1_tf" args="0 0 0 0 0 0 map $(arg frame_id_odom) 60"/> -->
      <node pkg="tf" type="static_transform_publisher" name="camera_tf" args="0.13 0 0.31 0 0 0 $(arg frame_id_robot_base) $(arg frame_id_camera_link) 60"/>

      <node pkg="potbot_example" type="twist_to_odometry" name="twist_to_odometry">
         <param name="publish_tf"                value="true"/>
         <param name="frame_id_odom"             value="$(arg frame_id_odom)"/>
         <param name="frame_id_robot_base"       value="$(arg frame_id_robot_base)"/>
         <remap from="rover_odo"                 to="/noise_twist"/>
         <remap from="odom"                      to="$(arg topic_odom)"/>
      </node>

      <node pkg="vmcl" name="sensor_filter" type="sensor_filter" output="screen">
         <remap from="odom"               to="/robot1/odom"/>
      </node>
   </group>

   <group if="$(arg use_rviz)">
      <node pkg="rviz" type="rviz" name="rviz_vmcl" required="true" args="-d $(arg file_rviz)"/>
   </group>

</launch>
